materialized view: make flow-control maximum delay configurable

piodul opened the pull request:
This pull request is continuation of scylladb/scylladb#20688 - contents of the main commit are the same, the only change is the additional commit with a test.

Until this patch, the materialized view flow-control algorithm (https://www.scylladb.com/2018/12/04/worry-free-ingestion-flow-control/) used a constant delay_limit_us hard-coded to one second, which means that when the size of view-update backlog reached the maximum (10% of memory), we delay every request by an additional second - while smaller amounts of backlog will result in smaller delays.

This hard-coded one maximum second delay was considered *huge* - it will slow down a client with concurrency 1000 to just 1000 requests per second - but we already saw some workloads where it was not enough - such as a test workload running very slow reads at high concurrency on a slow machine, where a latency of over one second was expected for each read, so adding a one second latecy for writes wasn't having any noticable affect on slowing down the client.

So this patch replaces the hard-coded default with a live-updateable configuration parameter, `view_flow_control_delay_limit_in_ms`, which defaults to 1000ms as before.

Another useful way in which the new `view_flow_control_delay_limit_in_ms` can be used is to set it to 0. In that case, the view-update flow control always adds zero delay, and in effect - does absolutely nothing. This setting can be used in emergency situations where it is suspected that the MV flow control is not behaving properly, and the user wants to disable it.

The new parameter's help string mentions both these use cases of the parameter.

Fixes #18187

This is new functionality, no need to backport to any open source release.

nyh commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -76,3 +81,112 @@ async def test_gossip_same_backlog(manager: ManagerClient) -> None:
+        else:
+            assert current_total_throttled_writes == previous_total_throttled_writes + 1
+
+        computed_delay = current_computed_delay - previous_computed_delay
```
By the way, there is also a metric last_mv_flow_control_delay - maybe it will be easier to use than the one that measures the total?

piodul replied:
Yes, using `last_mv_flow_control_delay` here should result in a bit less code. I absent-mindedly switched from it to `throttled_base_writes_total` when I introduced the `mv_flow_control_delay_total` check.

piodul replied:
Actually, I tried it and it the test started to fail. IIRC the calculated last delay was zero. I didn't dig into why this happened, but my theory is that there might have been some other writes happening which do not have the backpressure applied (different replica set?) and might have reset this counter to zero.

I think that using `view_flow_control_delay_limit_in_ms` will be more reliable here anyway.

piodul commented:
Sadly, the test doesn't work yet - it passes, but very rarely (or maybe not at all right now because I was tweaking it, but I did see it pass and it indeed verified that the delays were correct - it could be observed in the logs). When the test fails, it fails because the second write after the first, big one is not delayed (of course, it happens in the loop iteration when `delay_limit = 500`). Tracing shows that the node decides _not_ to delay the second write because the backlog that is sees is not big enough.

I don't think that the failure is caused by a bug in the first patch but rather by some nuances of how view backlog propagation works that I don't understand at the moment.

I will investigate the test further tomorrow.

nyh commented:
If you do just two writes and the second one doesn't get any delay at all, I can think of two possibilities:
1. Maybe your sever has two cpus and you got two requests to two different CPUs. Try starting your server with one CPU (sadly the test framework you are using is forcing you to create a server for your test, so might as well create one with just one CPU). Alternatively and perhaps more easily, just to more than two writes - if you have two cpus, the **third** write must reach one you already reached earlier :-)
2. Maybe the calculation of the delay, which takes the backlog size divided by 10% memory in **third power**, yields something too close to zero to be printed different from zero (maybe our printout loses precision or something?). You can reduce the amount of memory in the server (I don't remember if we have a parameter to reduce that 10% limit, maybe @wmitros remembers). Or perhaps just write a bit more?

piodul commented:
>     2. Maybe the calculation of the delay, which takes the backlog size divided by 10% memory in **third power**, yields something too close to zero to be printed different from zero (maybe our printout loses precision or something?). You can reduce the amount of memory in the server (I don't remember if we have a parameter to reduce that 10% limit, maybe @wmitros remembers). Or perhaps just write a bit more?

Yes, this was clearly missing from my test - this was handled by the `view_update_limit` error injection. Now, the test passes but only 50% of the time.

I also tried starting the nodes with only one shard, but it doesn't seem to change the situation much. The `pin_the_only_tablet` calls move all tablets to shard 0, so everything should already be happening within shard 0 (I assume that my driver is shard-aware).

piodul commented:
I opened an issue about a bug that gave me some hard time while debugging the test: https://github.com/scylladb/scylladb/issues/21672. After fixing it, I managed to improve the test so that it passes consistently.

The fix will influence behavior of view backpressure in a potentially significant way, so I think it deserves its own PR and would rather not include it here. I'll try another approach for this test, i.e. relying on view backlog propagation via gossip instead of via write response RPCs.

piodul commented:
> I opened an issue about a bug that gave me some hard time while debugging the test: #21672. After fixing it, I managed to improve the test so that it passes consistently.
> 
> The fix will influence behavior of view backpressure in a potentially significant way, so I think it deserves its own PR and would rather not include it here. I'll try another approach for this test, i.e. relying on view backlog propagation via gossip instead of via write response RPCs.

Actually, I realized that fixing #21672 isn't necessary. I also added an error injection to the test and misattributed the blame for breaking the test on #21672. I will send an updated PR soon.

piodul commented:
v2: debugged the test, now it passes reliably (on my machine, at least)

piodul commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -76,3 +81,112 @@ async def test_gossip_same_backlog(manager: ManagerClient) -> None:
+    await pin_the_only_tablet(manager, "ks", "mv_cf_view", servers[1])
+
+    # pin_the_only_tablet moves everything to shard 0
+    shard = 0
```
I updated the comment, it now just mentions that we start the two nodes with smp=1.

piodul commented:
v2.1: reworded a comment in order not to be misleading

piodul commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -76,3 +81,112 @@ async def test_gossip_same_backlog(manager: ManagerClient) -> None:
+        await manager.server_update_config(servers[0].server_id, "view_flow_control_delay_limit_in_ms", delay_limit)
+
+        # Make sure that the next view update will hang
+        await manager.api.enable_injection(servers[0].ip_addr, "never_finish_remote_view_updates", one_shot=False)
```
You're right, this comment is incorrect. Changed it.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13356/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13356/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_backlog
#### Failed Tests (1/90):
- [test_configurable_mv_control_flow_delay.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13356/testReport/junit/topology_custom/test_mv_backlog/test_configurable_mv_control_flow_delay_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_configurable_mv_control_flow_delay.debug.5)
#### Build Details:
- Duration: 45 min
- Builder: spider3.cloudius-systems.com

nyh commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -76,3 +81,106 @@ async def test_gossip_same_backlog(manager: ManagerClient) -> None:
+        computed_delays.append(computed_delay)
+
+        previous_computed_delay = current_computed_delay
+        previous_total_throttled_writes = current_total_throttled_writes
```
It's not clear to me why you write to `previous_*` the current value before unpausing the view update.
Wouldn't it be more reliable to fetch the `previous_*` in each iteration of the loop just before sending the write, and not in the middle of the previous iteration like you do here?

nyh replied:
Also there's another thing I'm wondering about. You're only doing here one big write (not a lot of small writes). How can we be confident that when the coordinator (by the way, I think we don't know which of the two nodes this is) wants to return success, it will already know the base replica (which might not be the coordinator itself!) has a high view backlog? Can't there be some delay until this base-replica statistic is known to the coordinator?

piodul replied:
> It's not clear to me why you write to `previous_*` the current value before unpausing the view update. Wouldn't it be more reliable to fetch the `previous_*` in each iteration of the loop just before sending the write, and not in the middle of the previous iteration like you do here?

Yeah, I get your point. It will be much more readable this way, i.e. easier for the reviewers/future readers of the test code to convince themselves that the metrics-related assertions work as intended.

piodul commented:
> #### Failed Tests (1/90):
> 
>     * [test_configurable_mv_control_flow_delay.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13356/testReport/junit/topology_custom/test_mv_backlog/test_configurable_mv_control_flow_delay_debug_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_configurable_mv_control_flow_delay.debug.5)

```
assert (26.9285 / 134.6425) > 0.9
 +  where 26.9285 = min([134.642, 134.642, 134.6425, 26.9285])
 +  and   134.6425 = max([134.642, 134.642, 134.6425, 26.9285])
```

`134.6425 / 26.9285 ~= 5`, which is the ratio between the two last tested values of the `view_flow_control_delay_limit_in_ms` parameter (2000 and 10000). It looks like the `server_update_config` function triggers the SIGHUP signal but doesn't wait until config is really reloaded.

nyh commented:
> `134.6425 / 26.9285 ~= 5`, which is the ratio between the two last tested values of the `view_flow_control_delay_limit_in_ms` parameter (2000 and 10000). It looks like the `server_update_config` function triggers the SIGHUP signal but doesn't wait until config is really reloaded.

Nice catch.

In my opinion, updating the config should **not** have used files or SIGHUP - it should have been done using the CQL interface for changing the configuration. It's significantly faster and - as you noticed - more guaranteed to work correctly. This is what we did in cqlpy test/cqlpy/util.py, config_value_context).
If you do want to do it through files, one way to fix this is to add a loop that verifies the new configuration (read system.config, using the CQL interface) and don't return until that's done.

piodul commented:
> In my opinion, updating the config should **not** have used files or SIGHUP - it should have been done using the CQL interface for changing the configuration. It's significantly faster and - as you noticed - more guaranteed to work correctly. This is what we did in cqlpy test/cqlpy/util.py, config_value_context).

Oh, that's really cool - I didn't know that we support changing live-updateable configuration parameters via CQL. Will check it out, thanks!

piodul commented:
> If you do want to do it through files, one way to fix this is to add a loop that verifies the new configuration (read system.config, using the CQL interface) and don't return until that's done.

This approach sounds most reasonable to me at the moment. I'd rather not spend time right now on changing test.py so that it uses `system.config` instead of config files, I expect it might not be that trivial.

piodul commented:
v3:
- Addressed Nadav's comment about measuring metrics right before and right after the base table write(s)
- Changed the test to use `system.config` instead of reloading Scylla configuration via SIGHUP - to make sure that the configuration is reloaded before the test proceeds to the part that expects the new value of the parameter,
- Now I'm doing one large write followed with a small one, instead of just the large write. This was made necessary because of the fact that I am now using `system.config` to change the delay parameter, which requires performing a __write__ to that table, and that affected the backlog calculation.

It turns out that the backlog returned in a replica response might not affect the write coordinator of that write, but only __the next write__ (I mentioned it in https://github.com/scylladb/scylladb/pull/21647#discussion_r1859174890).
- The test used to work previously because it submitted only one write in each operation, each time with the same size. Additionally, in the first iteration the delay limit would be set to 0. Because of those two facts, the write from the first operation would succeed because we expected it not to be delayed, however that was because it actually used an empty backlog for delay calculation! Other iterations would pass because it would take into the account the backlog of the write _from the previous operation_ - because all writes had the same size, it would look just as if the delay came from the same write.
- The test now requires two writes because the write to `system.config` is done when no view updates are in-flight on the node, so the backlog is calculated as 0 and then cached in the aforementioned coordinator cache. Now, we need the first big write to populate this cache and only the second one is affected by backlog in the cache and sets the delay metric.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13503/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13503/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_backlog
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/445/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5072/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1964/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13088/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13503/)
#### Build Details:
- Duration: 8 hr 37 min
- Builder: i-0603063e94862b0d8 (m5ad.8xlarge)

piodul commented:
@nyh @mitros review ping

wmitros commented on db/config.cc:
```diff
@@ -1179,6 +1179,11 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , topology_barrier_stall_detector_threshold_seconds(this, "topology_barrier_stall_detector_threshold_seconds", value_status::Used, 2, "Report sites blocking topology barrier if it takes longer than this.")
     , enable_tablets(this, "enable_tablets", value_status::Used, false, "Enable tablets for newly created keyspaces.")
+    , view_flow_control_delay_limit_in_ms(this, "view_flow_control_delay_limit_in_ms", liveness::LiveUpdate, value_status::Used, 1000,
+        "The maximal amount of that materialized-view update flow control may delay responses "
```
Missing word after "of" (I guess it was "time")

piodul replied:
Added "time" after "of"

wmitros commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -76,3 +81,113 @@ async def test_gossip_same_backlog(manager: ManagerClient) -> None:
+        if delay_limit == 0:
+            assert after_total_throttled_writes == before_total_throttled_writes
+        else:
+            assert after_total_throttled_writes >= before_total_throttled_writes + 1
```
In the comment above we say that we make sure there's only one throttled write, but here we assert that there's at least one throttled write. We should update the comment or the assert so that they match

piodul replied:
Good point. However, instead of doing that I changed the test in a different way - I measure the count/delay before and after the second write only, instead of including both writes. This way, if there is some change to how view backpressure is propagated and the first write starts being throttled, too, the test won't break.

wmitros reviewed (commented):
Looks good, fortunately the test seems to pass now. I added 2 comments which are small changes but I think they're worth fixing

piodul commented:
v3.1:
- Added "time" after "of" in the description of the "view_flow_control_delay_limit_in_ms" option
- Changed the test to only measure the delay of the second write to the base table, not both, and assert that there is exactly one throttled write

nyh commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -76,3 +81,115 @@ async def test_gossip_same_backlog(manager: ManagerClient) -> None:
+#   memory, influencing the view update backlog as a result.
+#   Follow it with one more small write to make sure that it is delayed
+#   by the first write's backlog.
+# - Read the X metric in order to measure the calculated delay of last view update.
```
Did you plan to go back and modify this X in the comment to the real metric name? Or did you write it this way because you check several metrics?

piodul replied:
I don't remember, but it's quite probable. In any case, changed `X` to `scylla_storage_proxy_coordinator_mv_flow_control_delay_total`.

piodul commented:
v3.2: changed "X" to "scylla_storage_proxy_coordinator_mv_flow_control_delay_total" in the test comment

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13633/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13633/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_backlog
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/528/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13196/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5180/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2070/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13633/)
#### Build Details:
- Duration: 9 hr 3 min
- Builder: i-042e53038151db56e (m5ad.8xlarge)

mykaul commented:
@scylladb/scylla-maint  - please review for merge.

mykaul commented:
This is a bit odd:
![image](https://github.com/user-attachments/assets/a2d341cc-b7a0-4c6f-9eb4-edfcfbc08ba0)

piodul commented:
> This is a bit odd: ![image](https://private-user-images.githubusercontent.com/4655593/396046767-a2d341cc-b7a0-4c6f-9eb4-edfcfbc08ba0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNDM1ODksIm5iZiI6MTczNDM0MzI4OSwicGF0aCI6Ii80NjU1NTkzLzM5NjA0Njc2Ny1hMmQzNDFjYy1iN2EwLTRjNmYtOWViNC1lZGZjZmJjMDhiYTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMTAwMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWEzNWNmN2E4MmZkOGI2OGNlODk3NjY1ODJhM2Q5YWQ0YTAyNmVkYzRlMWRkNWIyYjIwZmEwOTYyYTk0ZjhmYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.yvggeh4EhCmR-XvVz9c6RT4FGv31LAEvxKxdLSzrRpo)

Perhaps this is because I re-requested review from Nadav after he already approved.

In any case, the latest version of this PR only fixes nitpicks that were given in the last review which came with a LGTM - so I believe this is good to be merged.

@nyh / @scylladb/scylla-maint - please merge
